File: implementations/elixir/lib/rhizomatic/ed25519.ex
defmodule Rhizomatic.Ed25519 do
  @moduledoc """
  Ed25519 for the delta layer (SPEC-1 §5, §5.1).

  * **Signing** delegates to OTP's `:crypto` — RFC 8032 signing is
    deterministic and identical under every acceptance criterion.
  * **Verification** implements SPEC-1 §5.1's five-check STRICT criterion
    explicitly, in pure-Elixir Edwards arithmetic — never a library's default
    verifier, because library notions of "strict" vary:

      1. canonical scalar: `S < L` (little-endian);
      2. canonical encoding of `A` (decompress → recompress reproduces bytes);
      3. canonical encoding of `R` (same check);
      4. no small-order components: `[8]A ≠ 𝒪` and `[8]R ≠ 𝒪`;
      5. cofactorless equation, exactly: `[S]B = R + [k]A`,
         `k = SHA-512(R ‖ A ‖ M) mod L`.

  Correctness over speed: this runs in tests.
  """

  import Bitwise

  @p (1 <<< 255) - 19
  @l (1 <<< 252) + 27_742_317_777_372_353_535_851_937_790_883_648_493
  @d 37_095_705_934_669_439_343_138_083_508_754_565_189_542_113_879_843_219_016_388_785_533_085_940_283_555

  @bx 15_112_221_349_535_400_772_501_151_409_588_531_511_454_012_693_041_857_206_046_113_283_949_847_762_202
  @by 46_316_835_694_926_478_169_428_394_003_475_163_141_307_993_866_256_225_615_783_033_603_165_251_855_960

  # extended coordinates {X, Y, Z, T}; identity is (0 : 1 : 1 : 0)
  @identity {0, 1, 1, 0}

  # ------------------------------------------------------------------ keys

  @doc "Derive the 32-byte public key from a 32-byte seed (RFC 8032)."
  @spec public_key(binary()) :: binary()
  def public_key(<<seed::binary-32>>) do
    {pub, _} = :crypto.generate_key(:eddsa, :ed25519, seed)
    pub
  end

  @doc "Deterministic RFC 8032 signature (64 bytes) over `msg` by `seed`."
  @spec sign(binary(), binary()) :: binary()
  def sign(msg, <<seed::binary-32>>) do
    :crypto.sign(:eddsa, :none, msg, [seed, :ed25519])
  end

  # ---------------------------------------------------- strict verification

  @doc """
  SPEC-1 §5.1 strict verification of a 64-byte signature over `msg` against
  a 32-byte public key. Returns `true` iff all five checks pass.
  """
  @spec verify_strict(binary(), binary(), binary()) :: boolean()
  def verify_strict(<<r_bytes::binary-32, s_bytes::binary-32>>, msg, <<a_bytes::binary-32>>) do
    s = :binary.decode_unsigned(s_bytes, :little)

    with true <- s < @l,
         {:ok, a} <- decompress_canonical(a_bytes),
         {:ok, r} <- decompress_canonical(r_bytes),
         false <- small_order?(a),
         false <- small_order?(r) do
      k =
        :crypto.hash(:sha512, r_bytes <> a_bytes <> msg)
        |> :binary.decode_unsigned(:little)
        |> rem(@l)

      # [S]B = R + [k]A, exactly — no cofactor multiplication on either side
      point_equal(scalar_mult(s, basepoint()), point_add(r, scalar_mult(k, a)))
    else
      _ -> false
    end
  end

  def verify_strict(_, _, _), do: false

  # ----------------------------------------------------------- point codec

  # Canonical decode: the 32 bytes must decompress to a curve point whose
  # re-compression reproduces the identical bytes. Rejects y >= p and a set
  # sign bit when x = 0.
  defp decompress_canonical(<<bytes::binary-32>>) do
    n = :binary.decode_unsigned(bytes, :little)
    sign = n >>> 255 &&& 1
    y = n &&& (1 <<< 255) - 1

    cond do
      y >= @p ->
        :error

      true ->
        case recover_x(y) do
          {:ok, x} ->
            cond do
              x == 0 and sign == 1 -> :error
              (x &&& 1) != sign -> {:ok, to_extended(@p - x, y)}
              true -> {:ok, to_extended(x, y)}
            end

          :error ->
            :error
        end
    end
  end

  # x² = (y² - 1) / (d·y² + 1)  (twisted Edwards, a = -1)
  defp recover_x(y) do
    y2 = mulm(y, y)
    u = subm(y2, 1)
    v = addm(mulm(@d, y2), 1)
    # candidate root: (u/v)^((p+3)/8) = u · v³ · (u · v⁷)^((p-5)/8)
    v3 = mulm(mulm(v, v), v)
    v7 = mulm(mulm(v3, v3), v)
    x = mulm(mulm(u, v3), powm(mulm(u, v7), div(@p - 5, 8)))
    vx2 = mulm(v, mulm(x, x))

    cond do
      vx2 == mod(u) -> {:ok, x}
      vx2 == mod(-u) -> {:ok, mulm(x, powm(2, div(@p - 1, 4)))}
      true -> :error
    end
  end

  defp to_extended(x, y), do: {x, y, 1, mulm(x, y)}

  # ------------------------------------------------------ point arithmetic

  defp basepoint, do: to_extended(@bx, @by)

  # complete addition on twisted Edwards a=-1, extended coordinates
  defp point_add({x1, y1, z1, t1}, {x2, y2, z2, t2}) do
    a = mulm(subm(y1, x1), subm(y2, x2))
    b = mulm(addm(y1, x1), addm(y2, x2))
    c = mulm(mulm(2, @d), mulm(t1, t2))
    dd = mulm(2, mulm(z1, z2))
    e = subm(b, a)
    f = subm(dd, c)
    g = addm(dd, c)
    h = addm(b, a)
    {mulm(e, f), mulm(g, h), mulm(f, g), mulm(e, h)}
  end

  defp scalar_mult(k, point) when k >= 0 do
    do_scalar_mult(k, point, @identity)
  end

  defp do_scalar_mult(0, _p, acc), do: acc

  defp do_scalar_mult(k, p, acc) do
    acc = if (k &&& 1) == 1, do: point_add(acc, p), else: acc
    do_scalar_mult(k >>> 1, point_add(p, p), acc)
  end

  defp point_equal({x1, y1, z1, _}, {x2, y2, z2, _}) do
    mod(x1 * z2 - x2 * z1) == 0 and mod(y1 * z2 - y2 * z1) == 0
  end

  defp small_order?(point) do
    point_equal(scalar_mult(8, point), @identity)
  end

  # ---------------------------------------------------------- field helpers

  defp mod(x), do: Integer.mod(x, @p)
  defp addm(a, b), do: mod(a + b)
  defp subm(a, b), do: mod(a - b)
  defp mulm(a, b), do: mod(a * b)

  defp powm(base, exp) do
    :crypto.mod_pow(base, exp, @p) |> :binary.decode_unsigned()
  end
end

File: implementations/haskell/src/Rhizomatic/Ed25519.hs
-- | Ed25519 (RFC 8032) over Integer field arithmetic — signing, key
-- derivation, and the SPEC-1 §5.1 **strict** acceptance criterion, hand-rolled
-- clause by clause. Never a library's default verifier: the five checks below
-- are the normative text (ERRATA D13), and library notions of "strict" vary.
module Rhizomatic.Ed25519 (derivePublicKey, sign, verifyStrict) where

import qualified Data.ByteString as B
import Data.Bits (shiftL, shiftR, testBit, (.&.), (.|.))
import Data.Word (Word8)
import Rhizomatic.Sha512 (sha512)

p :: Integer
p = 2 ^ (255 :: Int) - 19

l :: Integer
l = 2 ^ (252 :: Int) + 27742317777372353535851937790883648493

d :: Integer
d = (-121665 * inv 121666) `mod` p

-- Extended homogeneous coordinates (X, Y, Z, T) with x = X/Z, y = Y/Z, xy = T/Z.
type Point = (Integer, Integer, Integer, Integer)

basepoint :: Point
basepoint =
  let by = (4 * inv 5) `mod` p
      bx = recoverX by 0
   in (bx, by, 1, (bx * by) `mod` p)

identity :: Point
identity = (0, 1, 1, 0)

inv :: Integer -> Integer
inv a = powMod a (p - 2) p

powMod :: Integer -> Integer -> Integer -> Integer
powMod _ 0 _ = 1
powMod b e m =
  let half = powMod b (e `shiftR` 1) m
      sq = (half * half) `mod` m
   in if testBit e 0 then (sq * (b `mod` m)) `mod` m else sq

pointAdd :: Point -> Point -> Point
pointAdd (x1, y1, z1, t1) (x2, y2, z2, t2) =
  let a = ((y1 - x1) * (y2 - x2)) `mod` p
      b = ((y1 + x1) * (y2 + x2)) `mod` p
      c = (2 * t1 * t2 * d) `mod` p
      dd = (2 * z1 * z2) `mod` p
      e = b - a
      f = dd - c
      g = dd + c
      h = b + a
   in ((e * f) `mod` p, (g * h) `mod` p, (f * g) `mod` p, (e * h) `mod` p)

scalarMult :: Integer -> Point -> Point
scalarMult 0 _ = identity
scalarMult n pt =
  let half = scalarMult (n `shiftR` 1) pt
      doubled = pointAdd half half
   in if testBit n 0 then pointAdd doubled pt else doubled

pointEq :: Point -> Point -> Bool
pointEq (x1, y1, z1, _) (x2, y2, z2, _) =
  (x1 * z2 - x2 * z1) `mod` p == 0 && (y1 * z2 - y2 * z1) `mod` p == 0

-- | x from y and the sign bit; error for non-square (callers use 'decompress'
-- for untrusted input — this is only reached for known-good curve points).
recoverX :: Integer -> Integer -> Integer
recoverX y sign_ = case xCandidate y of
  Just x -> if x `mod` 2 == fromIntegral sign_ then x else p - x
  Nothing -> error "ed25519: not a curve point"

-- | The candidate root with even spelling chosen canonically; Nothing when
-- x^2 = u/v has no root.
xCandidate :: Integer -> Maybe Integer
xCandidate y =
  let u = (y * y - 1) `mod` p
      v = (d * y * y + 1) `mod` p
      xx = (u * inv v) `mod` p
      cand = powMod xx ((p + 3) `shiftR` 3) p
      cand' = if (cand * cand) `mod` p == xx then cand else (cand * powMod 2 ((p - 1) `shiftR` 2) p) `mod` p
   in if (cand' * cand') `mod` p == xx then Just cand' else Nothing

compress :: Point -> B.ByteString
compress pt@(_, _, z, _) =
  let zi = inv z
      (x', y', _, _) = pt
      x = (x' * zi) `mod` p
      y = (y' * zi) `mod` p
      enc = y .|. ((x .&. 1) `shiftL` 255)
   in leBytes 32 enc

-- | Strict decompression: the 32 bytes must decode to a curve point whose
-- re-compression reproduces the identical bytes (SPEC-1 §5.1 checks 2–3:
-- rejects y >= p and a set sign bit when x = 0).
decompress :: B.ByteString -> Maybe Point
decompress bs
  | B.length bs /= 32 = Nothing
  | otherwise =
      let n = leInteger bs
          y = n .&. (2 ^ (255 :: Int) - 1)
          sign_ = n `shiftR` 255
       in if y >= p
            then Nothing
            else case xCandidate y of
              Nothing -> Nothing
              Just cand ->
                let x = if cand `mod` 2 == sign_ then cand else p - cand
                    pt = (x, y, 1, (x * y) `mod` p)
                 in if compress pt == bs then Just pt else Nothing

clamp :: B.ByteString -> Integer
clamp h =
  let n = leInteger (B.take 32 h)
   in (n .&. (2 ^ (254 :: Int) - 8)) .|. 2 ^ (254 :: Int)

-- | Public key bytes from a 32-byte seed.
derivePublicKey :: B.ByteString -> B.ByteString
derivePublicKey seed = compress (scalarMult (clamp (sha512 seed)) basepoint)

-- | RFC 8032 deterministic signature (64 bytes) over @msg@ by @seed@.
sign :: B.ByteString -> B.ByteString -> B.ByteString
sign seed msg =
  let h = sha512 seed
      s = clamp h
      prefix = B.drop 32 h
      pub = compress (scalarMult s basepoint)
      r = leInteger (sha512 (prefix <> msg)) `mod` l
      rPoint = compress (scalarMult r basepoint)
      k = leInteger (sha512 (rPoint <> pub <> msg)) `mod` l
      sScalar = (r + k * s) `mod` l
   in rPoint <> leBytes 32 sScalar

-- | SPEC-1 §5.1: the five checks, in the spec's order. True iff all hold.
verifyStrict :: B.ByteString -> B.ByteString -> B.ByteString -> Bool
verifyStrict pubBytes msg sigBytes
  | B.length sigBytes /= 64 || B.length pubBytes /= 32 = False
  | otherwise =
      let rBytes = B.take 32 sigBytes
          sBytes = B.drop 32 sigBytes
          sScalar = leInteger sBytes
       in -- 1. canonical scalar
          sScalar < l
            && case (decompress pubBytes, decompress rBytes) of
              -- 2–3. canonical encodings of A and R
              (Just aPoint, Just rPoint) ->
                -- 4. no small-order components: [8]A ≠ O and [8]R ≠ O
                not (pointEq (scalarMult 8 aPoint) identity)
                  && not (pointEq (scalarMult 8 rPoint) identity)
                  -- 5. cofactorless equation: [S]B = R + [k]A
                  && let k = leInteger (sha512 (rBytes <> pubBytes <> msg)) `mod` l
                      in pointEq (scalarMult sScalar basepoint) (pointAdd rPoint (scalarMult k aPoint))
              _ -> False

leInteger :: B.ByteString -> Integer
leInteger = B.foldr (\w acc -> (acc `shiftL` 8) .|. fromIntegral w) 0

leBytes :: Int -> Integer -> B.ByteString
leBytes n x = B.pack [fromIntegral (x `shiftR` (8 * i)) :: Word8 | i <- [0 .. n - 1]]
